### PR TITLE
Add vm scale sets to azure dynamic inventory

### PR DESCRIPTION
**Pull request by paktek123**
##### SUMMARY
Adds support for running playbooks on VM Scale Sets VMs. Filter the scale sets with tags, resource groups and location.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
`contrib`

##### ANSIBLE VERSION
```
2.5.2
```

### CODE REVIEW

**Comment by yuwzho** [comment]
@yungezz

File: contrib/inventory/azure_rm.py
Comment (yungezz) [inline contrib/inventory/azure_rm.py]: will need port also to access those vm

File: contrib/inventory/azure_rm.py
Comment (yungezz) [inline contrib/inventory/azure_rm.py:875]: get publicip info for vmss if has load balancer

File: contrib/inventory/azure_rm.py
Comment (yungezz) [inline contrib/inventory/azure_rm.py:863]: get security group info from vnet

**Comment by yungezz** [comment]
hello @paktek123 , thanks for contribution, could you pls look at comments?

**Comment by paktek123** [comment]
yup looking, sorry for the delay

File: contrib/inventory/azure_rm.py
Comment (paktek123) [reply contrib/inventory/azure_rm.py]: Hi @yungezz , thanks for reviewing. Sorry I don't understand. Can you please elaborate? I assumed the port was passed when running the ansible-playbook command.

File: contrib/inventory/azure_rm.py
Comment (paktek123) [reply contrib/inventory/azure_rm.py:863]: I cannot seem to find a way to associate the network_interface with the subnet . For example here is an id:
`/subscriptions/redacted/resourceGroups/test/providers/Microsoft.Compute/virtualMachineScaleSets/testvmss/virtualMachines/0/networkInterfaces/testvmss-nic'`

The virtual machine profile inside the vmss, returns None for network security groups :(
Also the nic doesn't seem to be part of the resource group. 

Only solution I think is running:

`self._network_client.network_interfaces.list_virtual_machine_scale_set_network_interfaces(resource_group, name)`

I get the subnet id and from that get the name and list the security groups underneath. I'll work on that on my next commit.

File: contrib/inventory/azure_rm.py
Comment (paktek123) [reply contrib/inventory/azure_rm.py:875]: added public ip info for vmss from the ip configuration in next commit

File: contrib/inventory/azure_rm.py
Comment (yungezz) [reply contrib/inventory/azure_rm.py:863]: you can get subnet info from `self.compute_client.virtual_machine_scale_sets.get(self.resource_group, self.name)`

File: contrib/inventory/azure_rm.py
Comment (yungezz) [reply contrib/inventory/azure_rm.py]: when vmss has loadbalancer, and user want to access vmss instance with public ip, will need port in inbound nat rules.

Comment (paktek123) [reply contrib/inventory/azure_rm.py]: thanks, I understand now

**Comment by yungezz** [comment]
@paktek123 HI Neeran, the changes looks good and inventory by vmss is very useful. Could you pls update based on comments? pls let me know if any help need, I can fork your branch to help if permission granted.

**Comment by yungezz** [comment]
@paktek123  would you like to  grant me forked  repo permission to continue the pr?  thanks.

**Comment by paktek123** [comment]
Hi @yungezz 

Sorry havent had time to work on this lately. I have given you write access to the forked ansible repo. Please let me know if you need help thanks.

File: contrib/inventory/azure_rm.py
Comment (zikalino) [inline contrib/inventory/azure_rm.py]: I guess for consistency everything should be uppercase "Specific Virtual Machine Scale Set". Should "scaleset" be one word or two words? "scale set"

File: contrib/inventory/azure_rm.py
Comment (zikalino) [inline contrib/inventory/azure_rm.py]: Virtual -> virtual, should be lowercase as it's not a title.

File: contrib/inventory/azure_rm.py
Comment (zikalino) [reply contrib/inventory/azure_rm.py]: Ok, I checked, It should be "scale set"

File: contrib/inventory/azure_rm.py
Comment (zikalino) [inline contrib/inventory/azure_rm.py]: why this downgrade?

File: contrib/inventory/azure_rm.py
Comment (zikalino) [inline contrib/inventory/azure_rm.py]: why not just "return [vmss]" here? and not continue the loop?

**Review by zikalino — APPROVED** [review]
Nothing major really.
The solution looks pretty straightforward and good.
I hope i haven't missed anything as i don't know the details of this code.

File: contrib/inventory/azure_rm.py
Comment (yungezz) [reply contrib/inventory/azure_rm.py]: fixed

File: contrib/inventory/azure_rm.py
Comment (yungezz) [reply contrib/inventory/azure_rm.py]: not downgrade, but revert upgrading in previous commit

File: contrib/inventory/azure_rm.py
Comment (yungezz) [reply contrib/inventory/azure_rm.py]: fixed

**Comment by yungezz** [comment]
@jborean93 HI Jordan, pls help to review the pr.

File: contrib/inventory/azure_rm.py
Comment (joshuaclausen) [inline contrib/inventory/azure_rm.py]: I had to change this back to the "enable_automatic_updates" when I tested on Ansible 2.6.2, run via AWX.

Comment (yungezz) [reply contrib/inventory/azure_rm.py]: thanks for reporting the issue, will work on it.

**Comment by yungezz** [comment]
support getting powerstate. sample output:
```
{
  "vmssnolb_3": {
    "resource_group": "xxx",
    "id": "/subscriptions/xxxx/resourceGroups/xxx/providers/Microsoft.Compute/virtualMachineScaleSets/vmssnolb/virtualMachines/3",
    "public_ip_name": null,
    "powerstate": "running",
    "public_ip": null,
    "public_ip_id": null,
    "fqdn": null,
    "plan": null,
    "network_interface": "vmssnolbNic",
    "ansible_connection": "winrm",
    "image": {
      "publisher": "MicrosoftWindowsServer",
      "offer": "WindowsServer",
      "sku": "2016-Datacenter",
      "version": "2016.127.20180613"
    },
    "type": "Microsoft.Compute/virtualMachineScaleSets/virtualMachines",
    "provisioning_state": "Succeeded",
    "tags": null,
    "computer_name": "vmssnolbq000003",
    "private_id": "10.0.0.7",
    "mac_address": null,
    "windows_timezone": null,
    "os_disk": {
      "operating_system_type": "Windows",
      "name": "vmssnolb_vmssnolb_3_OsDisk_1_6xxxxc0"
    },
    "network_interface_id": "/subscriptions/xxxx/resourceGroups/yungez-demo/providers/Microsoft.Compute/virtualMachineScaleSets/vmssnolb/virtualMachines/3/networkInterfaces/vmssnolbNic",
    "virtual_machine_size": "Standard_DS1_v2",
    "private_ip_alloc_method": "Dynamic",
    "private_ip": null,
    "windows_rm": null,
    "location": "eastus",
    "public_ip_alloc_method": null,
    "windows_auto_updates_enabled": true,
    "ansible_host": "10.0.0.7",
    "name": "vmssnolb_3"
  }
}
```

**Comment by yungezz** [comment]
ready_for_review

**Comment by yungezz** [comment]
@jborean93  @nitzmahone

**Review by yuwzho — APPROVED** [review]

**Comment by yuwzho** [comment]
shipit

**Comment by Fred-sun** [comment]
@paktek123 Thanks for your contribution, Could you help resolve the conflicting files?

**Comment by yungezz** [comment]
@Fred-sun  I'll resolve conflicts then

**Comment by Fred-sun** [comment]
Thanks  @yungezz, After you resolve conflicting, I will push others to review again.

**Comment by yungezz** [comment]
ready_for_review

**Comment by Fred-sun** [comment]
@joshuaclausen @yuwzho @zikalino could you help review this PR? Thanks!

**Comment by Fred-sun** [comment]
@joshuaclausen @yuwzho @zikalino could you help review this PR again? Thanks!

**Comment by Fred-sun** [comment]
@joshuaclausen @yuwzho @zikalino could you help review this PR again? Thanks!

**Comment by joshuaclausen** [comment]
If a vmss machine doesn't have a public IP address I get the following error:
TypeError: <azure.mgmt.network.v2017_06_01.models.public_ip_address.PublicIPAddress object at 0x7f2aa1dad0d0> is not JSON serializable

**Comment by yungezz** [comment]
@joshuaclausen thanks for reporting the issue. I didn't meet the error when working with vmss  with private ip only. Will check it and update here.

**Comment by joshuaclausen** [comment]
The problem with the public IP address is that the Azure API returns an object for "public_ip_address", rather than the actual IP address (yes, it's different than with the "private_ip_address".  So, if you append ".ip_address" to the end of line 727, everything will then work as expected:

vm.public_ip_address = nic_config.get("public_ip_address", None).ip_address

This shows the structure of the public_ip object returned that errors out as I had seen:
https://docs.microsoft.com/en-us/python/api/azure-mgmt-network/azure.mgmt.network.v2017_06_01.models.PublicIPAddress?view=azure-python

**Comment by joshuaclausen** [comment]
I got around the error by appending ".ip_address" to the end of the following line:
vm.public_ip_address = nic_config.get("public_ip_address", None).ip_address

However, the final host_vars['public_ip'] never had the public IP assigned, so it was "null".  After some more digging, it seems no public IPs come back in the function "_get_vmss_vm_instances".  When I print out the contents of nic_config.get("public_ip_address") as used within that function, I get "None" values for the public IP.  The following is what I get for all the nics in that function:

{'zones': None, 'dns_settings': None, 'name': None, 'tags': None, 'public_ip_address_version': None, 'public_ip_allocation_method': None, 'resource_guid': None, 'provisioning_state': None, 'ip_address': None, 'etag': None, 'location': None, 'ip_configuration': None, 'idle_timeout_in_minutes': None, 'type': None, 'id': u'/subscriptions/..../resourceGroups/.../providers/Microsoft.Compute/virtualMachineScaleSets/.../virtualMachines/11/networkInterfaces/primary-nic-cfg/ipConfigurations/primary-nic-ip-cfg/publicIPAddresses/...'}

I can confirm the machines in the vmss do indeed have a public IP, and I can connect to them on it.  I'll keep poking, but it's almost as though we're not getting any nic config data that includes the public IP.  I'm rather new to Ansible and Python, so forgive any noob mistakes ;)

**Comment by Fred-sun** [comment]
@yungezz do you have any updated about this PR? Thanks!

**Comment by moh-abk** [comment]
Any updates on this?

**Comment by yungezz** [comment]
@moh-abk  I was on something else recently. I will work on next week. For your information, there's a new inventory with batch api https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/azure_rm.py, you may want to have a try if urgent need.

**Comment by Fred-sun** [comment]
@yungezz

**Comment by Fred-sun** [comment]
@yungezz

**Comment by Fred-sun** [comment]
@yungezz

**Comment by Fred-sun** [comment]
@nitzmahone @jborean93 Could you help review this PR and push to forward？ Thanks!

**Comment by Fred-sun** [comment]
@paktek123 Thanks for contribution. This inventory script is in maintenance mode: only critical bug fixes but no new features. There's new Azure external inventory script at [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/azure_rm.py,) with better performance and latest new features. Pls move your contribution there. Thanks!

**Comment by Fred-sun** [comment]
@paktek123 Thanks for your contribution, Are you still here? Do you want to continue with this idea of PR or do you want to create a PR instead of this one? If you create new, please share to us! Thanks!

**Comment by Fred-sun** [comment]
@yuwzho The contributor disappear, can we closed it and reopen it when there is a need? Please check, Thank you very much!

**Comment by Fred-sun** [comment]
@yungezz Please help take a review this PR when you're free,  push it to merge or closing it! Thanks a lot!

**Comment by Fred-sun** [comment]
@MyronFanQiu The changes have been approved. Can you help push forward the merger? Thank you very much!

**Comment by Fred-sun** [comment]
@yungezz Contributors disappear. What do you do with the PR? Continue to wait for approval? Thank you very much!

**Comment by Fred-sun** [comment]
@yungezz @zikalino How move the PR going ahead? It has been approved but the contributor disappear! Please check! Thank you very much!

**Comment by Fred-sun** [comment]
@jborean93 @nitzmahone  This PR has been approved, please help to review and merge. Thank you very much!

**Comment by Fred-sun** [comment]
Ping to merge

**Comment by pixelicous** [comment]
Can this be merged please?? We really need to use scale sets with dynamic inventories!

**Comment by pixelicous** [comment]
Please merge this!!!

**Comment by Fred-sun** [comment]
@pixelicous  There's a new inventory with batch api https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/azure_rm.py. 
Can you try it on? Whether it can meet your needs. Thank you very much!

**Comment by Fred-sun** [comment]
@pixelicous I checked the previous ansible version and the current ansible, both of which can filter VMSS through resource groups and tags. You merge this, what is it mainly applicable to?
```e.g.
    - name: Get virtualmachine scale set
      azure_rm_virtualmachinescaleset_facts:
        resource_group: "{{ resource_group }}"
        tags:
          - "{{ tags_name }}"

**Comment by Fred-sun** [comment]
@paktek123 This PR has been Open for a long time and has been approved. Please help to confirm whether it can be merged. Thank you very much!

**Comment by paktek123** [comment]
This can be merged from my stand point, thanks

**Comment by pixelicous** [comment]
@Fred-sun well i didnt open this request but it seems like its like the issue i am having, i couldnt gett vmss out from the inventory and then run actions on them, maybe it was on my end, i didnt get back to it for a long time now, so i cant recall if i already solved it as well

**Comment by Fred-sun** [comment]
@gavinfish @haiyuazhang Can you help review and promote the PR merger? Thank you very much！

**Comment by Fred-sun** [comment]
@paktek123 Thank you for taking the time to contribute to this PR. We will transfer ansible 's azure module related Issue and PR to azure collection (https://github.com/ansible-collections/azure/pulls), can you transfer the Issue to azure collection repo?

**Pull request closed without merging**
